src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
         {
             return (BoundExpression)Visit(node);
         }
+
+        public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.ConstantValueOpt, node.MethodOpt, node.ResultKind, left, right, type));
+        }
+
+        public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.LogicalOperator, node.TrueOperator, node.FalseOperator, node.ResultKind, left, right, type));
+        }
+
+        private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator, Func<BoundBinaryOperatorBase, BoundExpression, BoundExpression, TypeSymbol, BoundBinaryOperatorBase> nodeUpdater)
+        {
+            // Use an explicit stack to avoid blowing the managed stack when visiting deeply-recursive
+            // binary nodes
+            var stack = ArrayBuilder<BoundBinaryOperatorBase>.GetInstance();
+            BoundBinaryOperatorBase currentBinary;
+            BoundExpression child = binaryOperator;
+
+            while (true)
+            {
+                if (child is BoundBinaryOperatorBase childBinary)
+                {
+                    currentBinary = childBinary;
+                    child = childBinary.Left;
+                    stack.Push(currentBinary);
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            Debug.Assert(stack.Count > 0);
+            currentBinary = null;
+
+            while (stack.Count > 0)
+            {
+                var left = currentBinary;
```

Comment:
>var left = currentBinary; [](start = 16, length = 25)

This feels more complicated than it needs to be. On the first iteration ```currentBinary``` is null. So, the ```updateNode``` has to have conditional logic to special case this and that logic will be executed on every iteration. It is not clear what is the benefit having that local function at all. It is used in one place and its body can simply be inlined. Also, we know that at least one iteration will  be done, checking condition for the first iteration is redundant and can be avoided by switching to a loop with post condition. To summarize all this, consider the following refactoring:
```
BoundExpression child = (BoundExpression)rewriter.Visit(stack.Peek().Left);

do
{
    currentBinary = stack.Pop();

                var foundInfo = rewriter._updatedNullabilities.TryGetValue(currentBinary, out (NullabilityInfo Info, TypeSymbol Type) infoAndType);
                var right = (BoundExpression)rewriter.Visit(currentBinary.Right);
                currentBinary = nodeUpdater(currentBinary, child, right, foundInfo ? infoAndType.Type : currentBinary.Type);

                if (foundInfo)
                {
                    currentBinary.TopLevelNullability = infoAndType.Info;
                }

     child = currentBinary;
}
while (stack.Count > 0);
```